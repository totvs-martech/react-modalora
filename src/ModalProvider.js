import React from 'react'
import PropTypes from 'prop-types'
import ModalController from './ModalController'
import ModalPortal from './ModalPortal'
import DefaultContentWrapper from './DefaultContentWrapper'
import DefaultModalContainer from './DefaultModalContainer'
import DefaultModalComponent from './DefaultModalComponent'
import defaultOnClose from './defaultOnClose'
import defaultOnOpen from './defaultOnOpen'
import getModalRootElementOriginal from './getModalRootElement'

function ModalProvider ({
  children,
  modalRoot,
  content,
  ContentWrapper = DefaultContentWrapper,
  getModalRootElement = getModalRootElementOriginal,
  ModalComponent = DefaultModalComponent,
  ModalContainer = DefaultModalContainer,
  onClose = defaultOnClose,
  onOpen = defaultOnOpen
} = {}) {
  const modalRootElement = getModalRootElement(modalRoot)

  return (
    <ModalController ModalComponent={ModalComponent} ContentComponent={content}>
      {({ ContentComponent, openModal, closeModal }) => {
        if (!ContentComponent) {
          return (
            <>
              <ContentWrapper isOpen={false}>
                {children({ openModal, closeModal })}
              </ContentWrapper>
            </>
          )
        }

        return (
          <>
            <ContentWrapper isOpen>
              {children({ openModal, closeModal })}
            </ContentWrapper>

            <ModalPortal modalRoot={modalRootElement} onOpen={onOpen} onClose={onClose}>
              <ModalContainer>
                <ContentComponent />
              </ModalContainer>
            </ModalPortal>
          </>
        )
      }}
    </ModalController>
  )
}

/* eslint-disable react/no-unused-prop-types */
ModalProvider.propTypes = {
  children: PropTypes.func.isRequired,
  content: PropTypes.oneOfType([PropTypes.element, PropTypes.string, PropTypes.func]),
  ContentWrapper: PropTypes.oneOfType([PropTypes.element, PropTypes.string, PropTypes.func]),
  getModalRootElement: PropTypes.func,
  ModalComponent: PropTypes.oneOfType([PropTypes.element, PropTypes.string, PropTypes.func]),
  ModalContainer: PropTypes.oneOfType([PropTypes.element, PropTypes.string, PropTypes.func]),
  modalRoot: PropTypes.oneOfType([PropTypes.instanceOf(HTMLElement), PropTypes.string]),
  onClose: PropTypes.func,
  onOpen: PropTypes.func
}
/* eslint-enable */

export const ContentWrapper = DefaultContentWrapper
export const ModalComponent = DefaultModalComponent
export const ModalContainer = DefaultModalContainer
export const onClose = defaultOnClose
export const onOpen = defaultOnOpen

export default ModalProvider
