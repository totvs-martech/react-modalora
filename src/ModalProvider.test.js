/* eslint-disable react/prop-types */
import React from 'react'
import { fireEvent, render } from 'react-testing-library'
import ModalProvider from './ModalProvider'

describe('ModalProvider', function () {
  describe('basic usage', function () {
    it('should match snapshot', function () {
      render(
        <ModalProvider>
          {() => (
            <>
              Modal content
            </>
          )}
        </ModalProvider>
      )

      expect(document.body.innerHTML).toMatchSnapshot()
    })
  })

  describe('providing content', function () {
    it('should render snapshot', function () {
      const MockContent = () => <>content</>

      render(
        <ModalProvider content={MockContent}>
          {() => (
            <>
              Modal content
            </>
          )}
        </ModalProvider>
      )

      expect(document.body.innerHTML).toMatchSnapshot()
    })
  })

  describe('calling openModal', function () {
    it('should render snapshots', function () {
      const MockContent = () => <>content</>
      let openModalRef

      render(
        <ModalProvider>
          {({ openModal }) => {
            openModalRef = openModal

            return (
              <>
                Modal content
              </>
            )
          }}
        </ModalProvider>
      )

      expect(document.body.innerHTML).toMatchSnapshot('before openModal')

      openModalRef(MockContent)
      expect(document.body.innerHTML).toMatchSnapshot('after openModal')
    })

    it('should set body.style.overflow=hidden', () => {
      const MockContent = () => <>content</>
      let openModalRef

      render(
        <ModalProvider>
          {({ openModal }) => {
            openModalRef = openModal

            return (
              <>
                Modal content
              </>
            )
          }}
        </ModalProvider>
      )

      expect(document.body.style.overflow).not.toBe('hidden')

      openModalRef(MockContent)
      expect(document.body.style.overflow).toBe('hidden')
    })
  })

  describe('calling closeModal', function () {
    it('should render snapshots', function () {
      const MockContent = () => <>content</>
      let closeModalRef

      render(
        <ModalProvider content={MockContent}>
          {({ closeModal }) => {
            closeModalRef = closeModal

            return (
              <>
                Modal content
              </>
            )
          }}
        </ModalProvider>
      )

      expect(document.body.innerHTML).toMatchSnapshot('before closeModal')

      closeModalRef()
      expect(document.body.innerHTML).toMatchSnapshot('after closeModal')
    })

    it('should unset body.style.overflow=hidden', function () {
      const MockContent = () => <>content</>
      let closeModalRef

      render(
        <ModalProvider content={MockContent}>
          {({ closeModal }) => {
            closeModalRef = closeModal

            return (
              <>
                Modal content
              </>
            )
          }}
        </ModalProvider>
      )

      expect(document.body.style.overflow).toBe('hidden')

      closeModalRef()
      expect(document.body.style.overflow).not.toBe('hidden')
    })
  })

  describe('passing closeOnOverlayClick=true', function () {
    it('should render snapshots', function () {
      const MockContent = () => <>content</>

      const { getByTestId } = render(
        <ModalProvider closeOnOverlayClick content={MockContent}>
          {() => (
            <>
              Modal content
            </>
          )}
        </ModalProvider>
      )

      expect(document.body.innerHTML).toMatchSnapshot('before click overlay')

      fireEvent.click(getByTestId('overlay'))

      expect(document.body.innerHTML).toMatchSnapshot('after click overlay')
    })
  })

  describe('passing custom', function () {
    describe('ContentWrapper', function () {
      it('should match snapshots', () => {
        const MockContentWrapper = ({ children, isOpen }) => <div data-isopen={isOpen}>{children}</div>
        const MockContent = () => <>content</>
        let openModalRef

        render(
          <ModalProvider ContentWrapper={MockContentWrapper}>
            {({ openModal }) => {
              openModalRef = openModal

              return (
                <>
                  Modal content
                </>
              )
            }}
          </ModalProvider>
        )

        expect(document.body.innerHTML).toMatchSnapshot('with isOpen=false')

        openModalRef(MockContent)
        expect(document.body.innerHTML).toMatchSnapshot('with isOpen=true')
      })
    })

    describe('ModalContainer', function () {
      it('should match snapshot', function () {
        const MockModalContainer = ({ children }) => <div data-testid='modal-container'>{children}</div>
        const MockContent = () => <>content</>

        render(
          <ModalProvider ModalContainer={MockModalContainer} content={MockContent}>
            {({ openModal }) => (
              <>
                Modal content
              </>
            )}
          </ModalProvider>
        )

        expect(document.body.innerHTML).toMatchSnapshot()
      })
    })

    describe('ModalComponent', function () {
      it('should match snapshot', function () {
        const MockModalComponent = ({ children }) => <div data-testid='modal-component'>{children}</div>
        const MockContent = () => <>content</>

        render(
          <ModalProvider ModalComponent={MockModalComponent} content={MockContent}>
            {({ openModal }) => (
              <>
                Modal content
              </>
            )}
          </ModalProvider>
        )

        expect(document.body.innerHTML).toMatchSnapshot()
      })
    })

    describe('modalRoot', function () {
      it('should pass modalRoot to getModalRootElement', () => {
        const mockGetModalRootElement = jest.fn(() => document.body)
        const mockModalRoot = 'modal-root'

        render(
          <ModalProvider getModalRootElement={mockGetModalRootElement} modalRoot={mockModalRoot}>
            {({ openModal }) => (
              <>
                Modal content
              </>
            )}
          </ModalProvider>
        )

        expect(mockGetModalRootElement).toHaveBeenCalledWith(mockModalRoot)
      })
    })

    describe('onOpen', () => {
      it('should be called passing content', () => {
        const mockOnOpen = jest.fn()
        const MockContent = () => <>content</>

        render(
          <ModalProvider onOpen={mockOnOpen} content={MockContent}>
            {() => (
              <>
                Modal content
              </>
            )}
          </ModalProvider>
        )

        expect(mockOnOpen).toHaveBeenCalled()
      })

      it('should be called on calling openModal', () => {
        const mockOnOpen = jest.fn()
        const MockContent = () => <>content</>
        let openModalRef

        render(
          <ModalProvider onOpen={mockOnOpen}>
            {({ openModal }) => {
              openModalRef = openModal

              return (
                <>
                  Modal content
                </>
              )
            }}
          </ModalProvider>
        )

        expect(mockOnOpen).not.toHaveBeenCalled()
        openModalRef(MockContent)
        expect(mockOnOpen).toHaveBeenCalled()
      })
    })

    describe('onClose', () => {
      it('should be called', () => {
        const mockOnClose = jest.fn()
        const MockContent = () => <>content</>
        let closeModalRef

        render(
          <ModalProvider onClose={mockOnClose} content={MockContent}>
            {({ closeModal }) => {
              closeModalRef = closeModal

              return (
                <>
                  Modal content
                </>
              )
            }}
          </ModalProvider>
        )

        expect(mockOnClose).not.toHaveBeenCalled()
        closeModalRef()
        expect(mockOnClose).toHaveBeenCalled()
      })
    })
  })
})
