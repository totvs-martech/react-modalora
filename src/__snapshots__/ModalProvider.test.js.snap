// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalProvider basic usage should match snapshot 1`] = `

<div>
  Modal content
</div>

`;

exports[`ModalProvider calling closeModal should render snapshots: after closeModal 1`] = `

<div>
  Modal content
</div>

`;

exports[`ModalProvider calling closeModal should render snapshots: before closeModal 1`] = `

<div>
  Modal content
</div>
<div>
  <div style="align-items: flex-start; bottom: 0px; display: flex; justify-content: center; left: 0px; max-height: 100vh; max-width: 100vw; overflow: auto; position: fixed; right: 0px; top: 0px;">
    <div style="background-color: rgba(0, 0, 0, 0.2); bottom: 0px; left: 0px; position: fixed; right: 0px; top: 0px; transform: translate3d(0px, -1px, 0px); width: 100%; padding: 0px; margin: 0px; border: 0px;">
    </div>
    <div role="dialog"
         style="background-color: rgb(255, 255, 255); margin: 3rem auto; max-width: 100vw; padding: 1.5rem; transform: translate3d(0px, 1px, 0px); width: 80rem;"
    >
      content
    </div>
  </div>
</div>

`;

exports[`ModalProvider calling openModal should render snapshots: after openModal 1`] = `

<div>
  Modal content
</div>
<div>
  <div style="align-items: flex-start; bottom: 0px; display: flex; justify-content: center; left: 0px; max-height: 100vh; max-width: 100vw; overflow: auto; position: fixed; right: 0px; top: 0px;">
    <div style="background-color: rgba(0, 0, 0, 0.2); bottom: 0px; left: 0px; position: fixed; right: 0px; top: 0px; transform: translate3d(0px, -1px, 0px); width: 100%; padding: 0px; margin: 0px; border: 0px;">
    </div>
    <div role="dialog"
         style="background-color: rgb(255, 255, 255); margin: 3rem auto; max-width: 100vw; padding: 1.5rem; transform: translate3d(0px, 1px, 0px); width: 80rem;"
    >
      content
    </div>
  </div>
</div>

`;

exports[`ModalProvider calling openModal should render snapshots: before openModal 1`] = `

<div>
  Modal content
</div>

`;

exports[`ModalProvider passing custom ContentWrapper should match snapshots: with isOpen=false 1`] = `

<div>
  <div data-isopen="false">
    Modal content
  </div>
</div>

`;

exports[`ModalProvider passing custom ContentWrapper should match snapshots: with isOpen=true 1`] = `

<div>
  <div data-isopen="true">
    Modal content
  </div>
</div>
<div>
  <div style="align-items: flex-start; bottom: 0px; display: flex; justify-content: center; left: 0px; max-height: 100vh; max-width: 100vw; overflow: auto; position: fixed; right: 0px; top: 0px;">
    <div style="background-color: rgba(0, 0, 0, 0.2); bottom: 0px; left: 0px; position: fixed; right: 0px; top: 0px; transform: translate3d(0px, -1px, 0px); width: 100%; padding: 0px; margin: 0px; border: 0px;">
    </div>
    <div role="dialog"
         style="background-color: rgb(255, 255, 255); margin: 3rem auto; max-width: 100vw; padding: 1.5rem; transform: translate3d(0px, 1px, 0px); width: 80rem;"
    >
      content
    </div>
  </div>
</div>

`;

exports[`ModalProvider passing custom ModalComponent should match snapshot 1`] = `

<div>
  Modal content
</div>
<div>
  <div style="align-items: flex-start; bottom: 0px; display: flex; justify-content: center; left: 0px; max-height: 100vh; max-width: 100vw; overflow: auto; position: fixed; right: 0px; top: 0px;">
    <div style="background-color: rgba(0, 0, 0, 0.2); bottom: 0px; left: 0px; position: fixed; right: 0px; top: 0px; transform: translate3d(0px, -1px, 0px); width: 100%; padding: 0px; margin: 0px; border: 0px;">
    </div>
    <div data-testid="modal-component">
      content
    </div>
  </div>
</div>

`;

exports[`ModalProvider passing custom ModalContainer should match snapshot 1`] = `

<div>
  Modal content
</div>
<div>
  <div data-testid="modal-container">
    <div role="dialog"
         style="background-color: rgb(255, 255, 255); margin: 3rem auto; max-width: 100vw; padding: 1.5rem; transform: translate3d(0px, 1px, 0px); width: 80rem;"
    >
      content
    </div>
  </div>
</div>

`;

exports[`ModalProvider providing content should render snapshot 1`] = `

<div>
  Modal content
</div>
<div>
  <div style="align-items: flex-start; bottom: 0px; display: flex; justify-content: center; left: 0px; max-height: 100vh; max-width: 100vw; overflow: auto; position: fixed; right: 0px; top: 0px;">
    <div style="background-color: rgba(0, 0, 0, 0.2); bottom: 0px; left: 0px; position: fixed; right: 0px; top: 0px; transform: translate3d(0px, -1px, 0px); width: 100%; padding: 0px; margin: 0px; border: 0px;">
    </div>
    <div role="dialog"
         style="background-color: rgb(255, 255, 255); margin: 3rem auto; max-width: 100vw; padding: 1.5rem; transform: translate3d(0px, 1px, 0px); width: 80rem;"
    >
      content
    </div>
  </div>
</div>

`;
